Synced code written so far

emanueleferrario opened the pull request.

emanueleferrario committed:
Synced code written so far

pianas commented on com/edocti/jintro/Exercitiul1.java:
```diff
@@ -0,0 +1,95 @@
+
+public class Exercitiul1 {
+
+public static Scanner scan=new Scanner(System.in);
```
In general, please place all operators between 2 spaces:
`int a = 10;`
`Scanner scan = new Scanner(...)` and so on

pianas commented on com/edocti/jintro/Exercitiul1.java:
```diff
@@ -0,0 +1,95 @@
+	}
+
+
+}
```
Small thing: it's always good to put a new line at the end of every source file.

pianas commented on com/edocti/jintro/Exercitiul1.java:
```diff
@@ -0,0 +1,95 @@
+//-----
+public static void main(String[] args) 
+{	
+System.out.println("Insert dimension the 2 Arrays ");	
```
When there is no indentation, the code is hard to maintain. Please always indent your code.

pianas commented on com/edocti/jintro/Exercitiul1.java:
```diff
@@ -0,0 +1,95 @@
+System.out.println("Insert dimension the 2 Arrays ");	
+DimensionArray();
+
+System.out.println("Insert the First Array");	
```
You don't need to perform all this input scan. You can simply test your method with a bunch of arrays that you (the programmer) provide. In this case you basically create a few test cases for your method(s).

pianas commented on com/edocti/jintro/Exercitiul1.java:
```diff
@@ -0,0 +1,95 @@
+	System.out.println(vector1[i]);}       
+//---
+
+System.out.println("Insert the Second Array");
```
In case you want to perform this input scan, you better create a method dedicated to this activity (readArr). Otherwise you copy-paste-replace the same code for every array.

pianas commented on com/edocti/jintro/Exercitiul1.java:
```diff
@@ -0,0 +1,95 @@
+//-----
+
+System.out.println("The dimension of the Vector Result is:");
+p=dim*2;
```
I don't want the two arrays to have the same dimension (dim). They can have any possible dimension.

pianas commented on com/edocti/jintro/Exercitiul1.java:
```diff
@@ -0,0 +1,95 @@
+for (int k=dim;k<p;k++) {
+	 vectorresult[k]=vector2[k-dim];}
+
+for (int j=0;j<p;j++) {
```
While this works, your algorithm is not efficient. It needs a number of steps proportional to m * n (hence O(m * n)), where m is the size of a1 and n is the size of a2. There are better sorting algorithms, but in this case you don't even need to sort, because you can use the fact that the arrays are already sorted => there is an O(n) algorithm.

pianas commented on com/edocti/jintro/Exercitiul1.java:
```diff
@@ -0,0 +1,95 @@
+System.out.println(vectorresult[k]);
+
+}
+
```
Way too many newlines between methods and inside the same method. There are some rules here:
- inside a method you can place one newline if you want to separate parts inside the same method. 
- Between methods, in Java people usually place one newline. Python instead has in my opinion a better code style (PEP8) which tells you to put 2 newlines - I personally adopted it also for Java and C++ code => max 2 lines between methods (up to you whether it's 1 or 2).

pianas commented on com/edocti/jintro/Exercitiul1.java:
```diff
@@ -0,0 +1,95 @@
+// ------
+
+public static int DimensionArray() {    // Input: Dimensiunea Array
+	dim=scan.nextInt();
```
again, `dim = scan.nextInt()` - notice the two spaces

pianas commented on com/edocti/jintro/Exercitiul2.java:
```diff
@@ -0,0 +1,98 @@
+public static void main(String[] args) 
+{	
+	System.out.println("Insert dimension of the Array");	
+    dim=scan.nextInt();
```
This is miss-aligned with the other statements from the same indentation level.

pianas commented on com/edocti/jintro/Exercitiul2.java:
```diff
@@ -0,0 +1,98 @@
+	//------
+	
+	System.out.println("VectorResult is :");   
+	int[] vectorresult=new int[dim];             // As vreau sa pun acest bloc de cod (linii 37-53) intr-o functie separata
```
The requirement is to rotate the array in-place => don't create a new array, but simply rotate the original one. Btw, all these exercises are far too long, taking care of all input scanning and so on. Instead, you can try to make a 10-line method which does the job and you can call it with some hard-coded values (arrays) from the `main()` method.

pianas commented on com/edocti/jintro/Exercitiul2.java:
```diff
@@ -0,0 +1,98 @@
+	System.out.println(n);
+	}
+	else do { 
+		System.out.println("Number of positions for rotation must be smaller or equal to "+dim+", Give another number of positions");
```
Not necessarily: you can rotate an array of 10 elements 995 positions to the left. This means that you only rotate it 5 positions actually. In other words, `n = n % arr.length;`

emanueleferrario committed:
added 2 new files

emanueleferrario committed:
added 3 files 08.10.2017

emanueleferrario committed:
added 1 file - Exercitiul 5

pianas commented on com/edocti/jintro/Exercitiul1.java:
```diff
@@ -0,0 +1,95 @@
+DimensionArray();
+
+System.out.println("Insert the First Array");	
+int [] vector1=new int[dim];
```
Same thing: = is between two spaces

pianas commented on com/edocti/jintro/Exercitiul1.java:
```diff
@@ -0,0 +1,95 @@
+int [] vector1=new int[dim];
+vector1[0]=scan.nextInt();
+System.out.println(vector1[0]);
+for (int i=1;i<dim;i++) {
```
Same thing here; all binary operators are between 2 spaces

pianas commented on com/edocti/jintro/Exercitiul1.java:
```diff
@@ -0,0 +1,95 @@
+System.out.println(vector1[0]);
+for (int i=1;i<dim;i++) {
+	vector1[i]=scan.nextInt();
+	if (vector1[i]>=vector1[i-1]) {
```
Same here

pianas commented on com/edocti/jintro/Exercitiul1.java:
```diff
@@ -0,0 +1,95 @@
+for (int i=1;i<dim;i++) {
+	vector1[i]=scan.nextInt();
+	if (vector1[i]>=vector1[i-1]) {
+		vector1[i]=vector1[i];
```
... and here

pianas commented on com/edocti/jintro/Exercitiul1.java:
```diff
@@ -0,0 +1,95 @@
+	vector1[i]=scan.nextInt();
+	if (vector1[i]>=vector1[i-1]) {
+		vector1[i]=vector1[i];
+		System.out.print(vector1[i]+" ");}
```
... and here

pianas commented on com/edocti/jintro/Exercitiul1.java:
```diff
@@ -0,0 +1,95 @@
+		System.out.print(vector1[i]+" ");}
+	else do { 
+		System.out.println("\nThe number you enter must be bigger than the last one");
+		vector1[i]=scan.nextInt();
```
... and here

pianas commented on com/edocti/jintro/Exercitiul1.java:
```diff
@@ -0,0 +1,95 @@
+		System.out.print(vector1[i]+" ");}
+	while ( vector1[i]<vector1[i-1]) ;
+	}
+for (int i=0;i<dim;i++) {
```
general indentation problem

pianas commented on com/edocti/jintro/Exercitiul5.java:
```diff
@@ -0,0 +1,137 @@
+
+//MediumAge
+private static void displayMediumAge (String lines) {
+	String [] line=lines.split ("\n");
```
no space between `String` and `[]`, but two spaces around `=`:
`String[] lines = ...`

pianas commented on com/edocti/jintro/Exercitiul5.java:
```diff
@@ -0,0 +1,137 @@
+	double medium;
+		for (int i=0;i<=line.length-1;i++) {
+		sum=sum+getAge(line[i]);
+	    		}
```
indentation problem

pianas reviewed (changes requested):
Please correct the errors reported and commit again to this branch

emanueleferrario committed:
added 1 file- EXERCISE6

PLEASE CHECK ONLY EXERCISE 6
I completed all the three "cerinte".
The cod has two problems:
1) the cod throws an exception in case o lack of one note in the mother String "students", as I explained in my mail of yesterday (19.10): there is a conflict with the method Integer.parseInt(String ......) in case of String "null"
2) the "cerinta 3" is not working good. 
    It's woorking OK for the methods "getWorstNote" and "getBestNote", but is not working good for the methods "getWorstStudent" and "getBestStudent", there is some problems in these two methods
Thanks
Emanuele

pianas commented on com/edocti/jintro/Exercise6.java:
```diff
@@ -0,0 +1,211 @@
+	public static String getName(String student) {
+		String[] lines = student.split("\n");
+		if (lines == null || lines.length == 0) {
+			return "You must insert the students !";
```
Not a good return value; you either throw an exception (which we didn't learn yet) or simply return null

pianas commented on com/edocti/jintro/Exercise6.java:
```diff
@@ -0,0 +1,211 @@
+
+	// getName
+	public static String getName(String student) {
+		String[] lines = student.split("\n");
```
A good thing is to always check your preconditions. In this case, you must assume that your array has 6 number of columns (or 7, not really sure)

pianas commented on com/edocti/jintro/Exercise6.java:
```diff
@@ -0,0 +1,211 @@
+	public static String getSurname(String student) {
+		String[] lines = student.split("\n");
+		if (lines == null || lines.length == 0) {
+			return "You must insert the students !";
```
same thing: better return null instead of that

pianas commented on com/edocti/jintro/Exercise6.java:
```diff
@@ -0,0 +1,211 @@
+	public static int[] getAverageNote(String student) {
+		String[] lines = student.split("\n");
+		int[] averagenote = new int[lines.length];
+		for (int i = 0; i <= lines.length - 1; i++) {
```
This method is a bit too complex (cyclomatic complexity). It's better to break it in multiple, smaller methods.

pianas commented on com/edocti/jintro/Exercise6.java:
```diff
@@ -0,0 +1,211 @@
+	}
+
+	// getAverageNote
+	public static int[] getAverageNote(String student) {
```
the name `student` is misleading; it gives me the impression that it's a single student.

pianas commented on com/edocti/jintro/Exercise6.java:
```diff
@@ -0,0 +1,211 @@
+	}
+
+	// checkNotaLucrare1
+	public static boolean checkNotaLucrare1(String student) {
```
please use English everywhere.

pianas reviewed (changes requested):
Please fix based on suggestion and I'll merge. This branch is too big already

emanueleferrario committed:
modified Exercise6

Please check only Exercise6

emanueleferrario committed:
Exercise6 - FINAL VERSION

emanueleferrario committed:
Exercise6 - FINAL VERSION - CORRECTED

emanueleferrario committed:
Added 1 file - EXERCISE 7

Please check EXERCISE 7

pianas commented on com/edocti/jintro/Exercise7.java:
```diff
@@ -0,0 +1,84 @@
+
+import java.lang.Math;
+
+public class Exercise7 {
```
You implemented all these yourself. Not that it's not possible, but there are already NumberFormats and Locales that deal with this problem. This exercise can be implemented in max 15 lines of code (including class and main()) and I encourage you to do it also like that (keep this one and add a new implementation).
